o1vm/src/pickles/verifier.rs

```diff
@@ -0,0 +1,267 @@
+    let zeta_chal = ScalarChallenge(fq_sponge.challenge());
+    let (_, endo_r) = G::endos();
+    let zeta: G::ScalarField = zeta_chal.to_field(endo_r);
+    let omega = domain.d1.group_gen;
+    let zeta_omega = zeta * omega;
+
+    let column_eval = ColumnEval {
+        commitment: commitments,
+        zeta_eval: zeta_evaluations,
+        zeta_omega_eval: zeta_omega_evaluations,
+    };
+
+    // -- Absorb all commitments_and_evaluations
+    let fq_sponge_before_commitments_and_evaluations = fq_sponge.clone();
+    let mut fr_sponge = EFrSponge::new(G::sponge_params());
+    fr_sponge.absorb(&fq_sponge.digest());
+
+    for (zeta_eval, zeta_omega_eval) in zeta_evaluations
+        .scratch
+        .iter()
+        .zip(zeta_omega_evaluations.scratch.iter())
+    {
+        fr_sponge.absorb(zeta_eval);
+        fr_sponge.absorb(zeta_omega_eval);
+    }
+    fr_sponge.absorb(&zeta_evaluations.instruction_counter);
+    fr_sponge.absorb(&zeta_omega_evaluations.instruction_counter);
+    fr_sponge.absorb(&zeta_evaluations.error);
+    fr_sponge.absorb(&zeta_omega_evaluations.error);
+    fr_sponge.absorb_multiple(&zeta_evaluations.selector);
+    fr_sponge.absorb_multiple(&zeta_omega_evaluations.selector);
+    fr_sponge.absorb(&quotient_evaluations.zeta);
+    fr_sponge.absorb(&quotient_evaluations.zeta_omega);
+    // FIXME: Add selector evaluations (DONE) and quotient evaluations
+
+    // FIXME: use a proper Challenge structure
+    let challenges = BerkeleyChallenges {
+        alpha,
+        // No permutation argument for the moment
+        beta: G::ScalarField::zero(),
+        gamma: G::ScalarField::zero(),
+        // No lookup for the moment
+        joint_combiner: G::ScalarField::zero(),
+    };
+    let (_, endo_r) = G::endos();
+
+    let constants = Constants {
+        endo_coefficient: *endo_r,
+        mds: &G::sponge_params().mds,
+        zk_rows: 0,
+    };
+
+    let combined_expr =
+        Expr::combine_constraints(0..(constraints.len() as u32), constraints.clone());
+
+    // FIXME: Add these to the final check!!!!!
+
+    // FIXME: Fixup absorbs so they match in prover.rs
+
+    let quotient_eval_zeta = PolishToken::evaluate(
+        combined_expr.to_polish().as_slice(),
+        domain.d1,
+        zeta,
+        &column_eval,
+        &constants,
+        &challenges,
+    )
+    .unwrap_or_else(|_| panic!("Could not evaluate quotient polynomial at zeta"));
+
+    let quotient_eval_zeta_omega = PolishToken::evaluate(
+        combined_expr.to_polish().as_slice(),
+        domain.d1,
+        zeta_omega,
+        &column_eval,
+        &constants,
+        &challenges,
+    )
+    .unwrap_or_else(|_| panic!("Could not evaluate quotient polynomial at zeta_omega"));
+
+    // Check the actual quotient works. combined_expr(eval) [ == quotient_eval_*] = quotient(eval) [== Given by prover (new field) -- chunked] * vanishing_poly(eval) [== x^n - 1 == zeta^(d1.size()) - 1]
```

Comment:
Need to do this still.

Reply:
Unsure how to do this, see current code for the assert I'm assuming we want to run as some sort of verified assert.